src/storage-controller/src/lib.rs

```diff
@@ -569,61 +578,127 @@ where
+                    let source_collection = self
+                        .collections
+                        .get_mut(&ingestion_id)
+                        .expect("known to exist");
+                    match &mut source_collection.description {
+                        CollectionDescription {
+                            data_source: DataSource::Ingestion(ingestion_desc),
+                            ..
+                        } => {
+                            // DataSource::IngestionExport names the object it
+                            // wants to export, so we look up the output index
+                            // for that name.
+                            let output_index = ingestion_desc
+                                .desc
+                                .connection
+                                .output_idx_for_name(&external_reference)
+                                .ok_or(StorageError::MissingSubsourceReference {
+                                    ingestion_id,
+                                    reference: external_reference,
+                                })?;
+
+                            ingestion_desc.source_exports.insert(
+                                id,
+                                SourceExport {
+                                    output_index,
+                                    storage_metadata: (),
+                                },
+                            )
+                        }
+                        _ => unreachable!(
+                            "SourceExport must only refer to primary sources that already exist"
+                        ),
+                    };
+
+                    self.collections.insert(id, collection_state);
 
-                to_create.push((id, description));
+                    // Executing the source export doesn't do anything, ensure we execute the source instead.
+                    to_execute.remove(&id);
+                    to_execute.insert(ingestion_id);
+
+                    new_source_statistic_entries.insert(id);
+                }
+                DataSource::Other(DataSourceOther::TableWrites) => {
+                    debug!(desc = ?description, meta = ?metadata, "registering {} with persist table worker", id);
+                    table_registers.push((id, write, collection_state));
+                }
+                DataSource::Progress | DataSource::Other(DataSourceOther::Compute) => {
+                    debug!(desc = ?description, meta = ?metadata, "not registering {} with a controller persist worker", id);
+                    self.collections.insert(id, collection_state);
+                }
+                DataSource::Ingestion(_) => {
+                    debug!(desc = ?description, meta = ?metadata, "not registering {} with a controller persist worker", id);
+                    self.collections.insert(id, collection_state);
+                    new_source_statistic_entries.insert(id);
+                }
```

Comment:
Oy––this was a good comment. I think the original thing I did for `execute_collections` was too easy to misuse (the only thing that should ever be re-executed are ingestions). I refactored this to `run_ingestion` and it now only runs ingestions (since that is the only collection type that should ever be re-rendered).